### PR TITLE
Chapter 9

### PR DESCRIPTION
**Pull request by dmitryrotay**
Adding a base shape type and implementing planes.

### CODE REVIEW

File: src/Shapes.hs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            | otherwise = normalVector
+        overPoint = point `addVectorP` (normalVector' `multiplyVector` epsilon)
+    in Computations shape distance point overPoint eyeVector normalVector' isInside
+
+hit :: [Intersection] -> Maybe Intersection
+hit xs = 
+    let positiveIntersections = filter (\(Intersection _ t) -> t >= 0) xs
+    in case positiveIntersections of
+        [] -> Nothing
+        _ -> Just $ minimumBy
```
Comment (EduardSergeev) [inline src/Shapes.hs:136]: Could be written as:

```haskell
Just . minimumBy (comparing getDistance) $ positiveIntersections
```

File: src/Shapes.hs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            , getShapeTransform :: Transform
+            , getShapeInverseTransform :: Transform
+            , getShapeMaterial :: Material
+            } deriving (Show, Eq)
+
+data Intersection = Intersection { getShape :: Shape, getDistance :: Double }
+    deriving (Show, Eq)
+
+instance Ord Intersection where
+    compare i1 i2 = compare (getDistance i1) (getDistance i2)
```
Comment (EduardSergeev) [inline src/Shapes.hs:48]: Or just:

```haskell
compare = compare `on` getDistance
```

[on](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Function.html#v:on) is a useful combinator

File: src/World.hs [hunk 3]
```diff
@@ -42,20 +40,20 @@ defaultWorld =
...
+        sphere2' = setTransform sphere2  (scaling 0.5 0.5 0.5)
         light = PointLight (Point (-10) 10 (-10)) (Color 1 1 1)
     in World [sphere1', sphere2'] (Just light)
 
-intersectWorld :: World -> Ray -> [Intersection Sphere]
-intersectWorld (World objects _) ray =
-    sort $ concat [intersectionToList (object `intersect` ray) | object <- objects]
+intersectWorld :: World -> Ray -> [Intersection]
+intersectWorld (World shapes _) ray =
+    sort $ concat [shape `intersect` ray | shape <- shapes]
```
Comment (EduardSergeev) [inline src/World.hs:49]: Or:
```haskell
    sort $ flip intersect ray =<< shapes
```

File: src/Shapes.hs (hunk 1 above)
Comment (dmitryrotay) [reply src/Shapes.hs:136]: Had to throw away this part. I forgot to change it back to `minimum` after I re-introduced `Ord` instance for this data type. 
`comparing` seems like a cool shortcut, though. Thanks!

File: src/Shapes.hs (hunk 2 above)
Comment (dmitryrotay) [reply src/Shapes.hs:48]: Had to meditate on its type signature a bit. Yep, that looks nice.

File: src/World.hs (hunk 3 above)
Comment (dmitryrotay) [reply src/World.hs:49]: Okay. I guess every time I need `concat` it's makes sense to first think about the list monad...

Comment (EduardSergeev) [reply src/World.hs:49]: Pretty much :) `m >>= f` for list can be defined as: `concat . fmap f $ m` and list comprehension is just `fmap`.
With more practice you'll recognise these patterns automatically.
And `>>=` or `=<<` very often can simplify the code

**Pull request merged**